frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/services/JwtService.java

```diff
@@ -0,0 +1,94 @@
+package uk.gov.di.authentication.frontendapi.services;
+
+import com.nimbusds.jose.EncryptionMethod;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWEAlgorithm;
+import com.nimbusds.jose.JWEHeader;
+import com.nimbusds.jose.JWEObject;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.Payload;
+import com.nimbusds.jose.crypto.RSAEncrypter;
+import com.nimbusds.jose.crypto.impl.ECDSA;
+import com.nimbusds.jose.util.Base64URL;
+import com.nimbusds.jwt.EncryptedJWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.kms.model.KmsException;
+import software.amazon.awssdk.services.kms.model.SignRequest;
+import software.amazon.awssdk.services.kms.model.SignResponse;
+import software.amazon.awssdk.services.kms.model.SigningAlgorithmSpec;
+import uk.gov.di.authentication.shared.services.KmsConnectionService;
+
+import java.nio.charset.StandardCharsets;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+
+import static com.nimbusds.jose.crypto.impl.ECDSA.getSignatureByteArrayLength;
+
+public class JwtService {
+    private static final Logger LOG = LogManager.getLogger(JwtService.class);
+    private final KmsConnectionService kmsConnectionService;
+
+    public JwtService(KmsConnectionService kmsConnectionService) {
+        this.kmsConnectionService = kmsConnectionService;
+    }
+
+    public SignedJWT signJWT(
+            JWSAlgorithm signingAlgorithm, JWTClaimsSet jwtClaims, String keyAlias) {
+        LOG.info("Creating signed JWT");
+        JWSHeader jwsHeader = new JWSHeader(signingAlgorithm);
+        Base64URL encodedHeader = jwsHeader.toBase64URL();
+        Base64URL encodedClaims = Base64URL.encode(jwtClaims.toString());
+        String message = encodedHeader + "." + encodedClaims;
+        SignRequest signRequest =
+                SignRequest.builder()
+                        .message(SdkBytes.fromByteArray(message.getBytes(StandardCharsets.UTF_8)))
+                        .keyId(keyAlias)
+                        .signingAlgorithm(SigningAlgorithmSpec.ECDSA_SHA_256)
+                        .build();
+        try {
+            LOG.info("Signing JWT");
+            SignResponse signResult = kmsConnectionService.sign(signRequest);
+            LOG.info("JWT has been signed successfully");
+            String signature =
+                    Base64URL.encode(
+                                    ECDSA.transcodeSignatureToConcat(
+                                            signResult.signature().asByteArray(),
+                                            getSignatureByteArrayLength(signingAlgorithm)))
+                            .toString();
```

Comment:
Ah good catch! Would've led to some funny error logging

Reply:
Just to flag, if it fails to parse it, it will still say  `"JWT has been signed successfully"` then `"Failed to sign JWT"` - this will probably be naturally fixed when throwing more specific exceptions but just noting it in case.

Reply:
Good spot, I've fixed this by moving the parsedJWT into a variable and moving the log after the parse call is made and then returning. Means that we don't log the success message if we then fail to parse